torch/testing/asserts.py

```diff
@@ -0,0 +1,389 @@
+        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :attr:`~torch.Tensor.dtype`.
+        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :meth:`~torch.Tensor.stride`.
+
+    Raises:
+        AssertionError: If :attr:`a` and :attr:`b` do not have the same :attr:`~torch.Tensor.shape`.
+        AssertionError: If :attr:`check_device`, but :attr:`a` and :attr:`b` do not live in the same
+            :attr:`~torch.Tensor.device` memory.
+        AssertionError: If :attr:`check_dtype`, but :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`.
+        AssertionError: If :attr:`check_stride`, but :attr:`a` and :attr:`b` do not have the same
+            :meth:`~torch.Tensor.stride`.
+    """
+    msg_fmtstr = "The values for attribute '{}' do not match: {} != {}."
+
+    if a.shape != b.shape:
+        raise AssertionError(msg_fmtstr.format("shape", a.shape, b.shape))
+
+    if check_device and a.device != b.device:
+        raise AssertionError(msg_fmtstr.format("device", a.device, b.device))
+
+    if check_dtype and a.dtype != b.dtype:
+        raise AssertionError(msg_fmtstr.format("dtype", a.dtype, b.dtype))
+
+    if check_stride and a.stride() != b.stride():
+        raise AssertionError(msg_fmtstr.format("stride()", a.stride(), b.stride()))
+
+
+def _assert_values_equal(a: torch.Tensor, b: torch.Tensor):
+    """Asserts that the values of two tensors are bitwise equal.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+
+    Raises:
+         AssertionError: If the values of :attr:`a` and :attr:`b` are not bitwise equal.
+
+    .. seealso::
+
+        Internally :func:`torch.eq` is used to check for bitwise equality.
+    """
+    if not torch.all(torch.eq(a, b)):
+        raise AssertionError("ADDME")
+
+
+def _assert_values_allclose(
+    a: torch.Tensor, b: torch.Tensor, /, *, rtol: Optional[float] = None, atol: Optional[float] = None
+) -> None:
+    """Asserts that the values of two tensors are equal up to a desired precision.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        rtol (Optional[float]): Relative tolerance. If omitted, a default value based on the
+            :attr:`~torch.Tensor.dtype` is selected with the table from :func:`assert_tensors_allclose`.
+        atol (Optional[float]): Absolute tolerance. If omitted, a default value based on the
+            :attr:`~torch.Tensor.dtype` is selected with the table from :func:`assert_tensors_allclose`.
+
+    Raises:
+         AssertionError: If the values of :attr:`a` and :attr:`b` are equal up to the desired precision.
+
+    .. seealso::
+
+        Internally :func:`torch.allclose` is used to check for bitwise equality.
+    """
+    default_rtol, default_atol = _get_default_rtol_and_atol(a, b)
```

Comment:
Currently we require both rtol and atol be set if either is set:

https://github.com/pytorch/pytorch/blob/0c8f16622bf6f410b67adec08c15692d5c404225/torch/testing/_internal/common_utils.py#L1078

This was to avoid someone setting atol, for example, and then being surprised that rtol was not 0.

For simplicity I suggest keeping this restriction for now. If we think it's too onerous then we can remove it in the future. If we don't keep it, however, it may be challenging to add later if we decide it's a good idea.